[14.0][BACKPORT] project_forecast_line from 15.0 to 14.0

ntsirintanis opened the pull request:
Removed any relation to `project.project.stage`, as it is not available in 14.0

yankinmax committed:
[15.0][ADD] project_forecast_line

Commit:
[UPD] Update project_forecast_line.pot

OCA-git-bot committed:
[UPD] README.rst

gurneyalex committed:
project_forecast_line: improve documentation

gurneyalex committed:
project_forecast_line: improve documentation

yankinmax committed:
[15.0][FIX] project_forecast_line: take roles into account on calc

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
project_forecast_line 15.0.1.0.1

gurneyalex committed:
[FIX] project_forecast_line consolidated capacity

in some cases, we could have capacity consumed by a task which was not matching
a work capacity line in the period because of a faulty optimisation we were making
which skipped the creation of lines with a capacity of 0 -> then we had no line on which
to compute the negative consolidated capacity.

We remove the optimisation to fix this case and show the problematic periods
in the consolidated capacity graphs

OCA-git-bot committed:
project_forecast_line 15.0.1.0.2

yankinmax committed:
[15.0][IMP] project_forecast_line: setting to control consumption states

gurneyalex committed:
bump version after clicking on merge by mistake

Commit:
[UPD] Update project_forecast_line.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: project-15.0/project-15.0-project_forecast_line
Translate-URL: https://translation.odoo-community.org/projects/project-15-0/project-15-0-project_forecast_line/

yankinmax committed:
[15.0][FIX] project_forecast_line: add missing access rights

ntsirintanis committed:
[14.0][BACKPORT][WIP]project_forecast_line from 15.0 to 14.0

gurneyalex commented:
please add backports of
* #973 
* #974 
* #971 
* #962

gurneyalex commented:
Regarding project.stage, the equivalement feature on OCA is using project status. Can I suggest that you either replace odoo 15's project stage with OCA's 14.0 project status ? https://github.com/OCA/project/tree/14.0/project_status

ntsirintanis commented:
> Regarding project.stage, the equivalement feature on OCA is using project status. Can I suggest that you either replace odoo 15's project stage with OCA's 14.0 project status ? https://github.com/OCA/project/tree/14.0/project_status

I'll look into that, thanks for letting me know :+1:

ntsirintanis committed:
[14.0][IMP] project_forecast_line: replace project.project.stage with project.status

ntsirintanis commented on project_forecast_line/models/project_task.py:
```diff
@@ -0,0 +1,156 @@
+                sale_state = task.sale_line_id.state
+                if sale_state == "cancel":
+                    _logger.info("skip task %s: cancelled sale", task)
+                    continue
```
@gurneyalex I added this `continue` here, but it is not there in 15.0 version. In my understanding, it should be there. Am I missing something?

gurneyalex committed:
[FIX] project_forecast_line: error in consolidated capacity

During some operations, forecast lines are deleted and recreated. This
can lead to some hr.employee.forecast.role lines being deleted which
leaves the forecast lines on the same period without a related document
to store the consolidated capacity. Normally that line is recreated
shortly afterwards but the creation does not recompute the link between
the parent-less lines and the new one. This patches forces the
recomputation when new lines are created.

yankinmax committed:
Revert "[15.0][IMP] project_forecast_line: setting to control consumption states"

This reverts commit 82f5f363fe43f83c82a4a374314bf489fe12bb99.

yankinmax committed:
[FIX] project_forecast_line: give sudo rights on hr.leave to create forecast

ntsirintanis commented:
@gurneyalex after cherry-picking https://github.com/OCA/project/pull/979 , tests started failing. More specifically, the one test that was introduced by this PR, is failing. Which is strange, because it does not fail on local. Can you please help on this?

gurneyalex committed:
[FIX] project_forecast_line: handle lack of project status

In some cases you can have a project without a status set -> in this
case we don't want to generate forecast lines (and we don't want a crash
either)

gjotten commented:
> It looks like some users haven't signed our **C**ontributor **L**icense **A**greement, yet. You can read and sign our full Contributor License Agreement here: https://odoo-community.org/page/cla Here is a list of the users:
> 
> * @gjotten (login unknown in OCA database)

Oops, my bad, should be fixed soon since I sent it out just now.

gurneyalex committed:
fix the computation of the field and the tests

gurneyalex committed:
project_forecast_line: add demo data

add demo data to the module to help making demos

closes: #990

gurneyalex committed:
fix: check on employee main role computation

yankinmax committed:
[15.0][IMP] project_forecast_line:

Add computed fields to written fields to trigger forecast lines update,
Add hook for several UoM hours/days conversion

yankinmax committed:
[IMP] Reorder search filters

ntsirintanis committed:
[14.0][IMP][WIP] project_forecast_line: fix backport unit tests

ntsirintanis commented:
@thomaspaulb @NL66278 I suspect that the reason this one test is failing, is because runboat fails as well in a specific module:

https://runboat.odoo-community.org/api/v1/builds/b2c502af6-71bb-42da-84c8-003b8937bbc7/init-log

Not really sure how to fix this. Locally though the tests are fine.

@gjotten Please go ahead and test this functionally again, if you have time

thomaspaulb commented:
@ntsirintanis From the looks of the error message in the test run, it's a conflict between `project_task_code` and `project_timesheet_holidays`. As a first line of attack though, could you rebase your branch on the latest `14.0`? You seem to be ~70 commits behind and maybe there is already something in those last commits that solved it, and we don't need to go into it. 14.0 branch itself looks green.

yankinmax committed:
[15.0][UPD] project_forecast_line: update forecast lines even if employee is not active

gurneyalex committed:
refactor(project_forecast_line): computed field trigger

refactor the triggers for recomputation of tasks related to project.task
and sale.order.lines to use an override of the _write() method to catch
the computed fields changes.

remove overrides of analytic line creation / update

add a call to project.task::flush() in the test to force the recomputation

add a call to project.project::flush() after changing the project
stage to force the recomputation

remove the utils module

reduce mail in the tests (tracking_disable=1)

make prepare_forecast_lines private

gurneyalex committed:
[IMP] speed up for project task forecast

1. when only the remaining time of a project task is updated (e.g. when a
user timesheets) then use a quick update that will compute the new
forecast_hours on each line for the employee by applying a ratio
new_remaining_hours / old_remaining_hours

2. in the normal path of forecast.line::_update_forecast_lines we write
some values on the lines. This can trigger the recomputation of
employee_resource_forecast_line_id which takes a significant amount of
time -> avoid needless recomputation by avoiding a write of unchanged
values

gurneyalex committed:
fix: unlink forecast lines when parent model is unlinked

gurneyalex committed:
refactor: extract a mixin class

add a mixin class to meld in models which generate forecast lines

gurneyalex committed:
add some logging about forecast line creation / update

gurneyalex committed:
switch to forecast line update on employee role update

gurneyalex committed:
handle refusal of leaves after validation

in that case we must recreate the employee role forecast lines

gurneyalex committed:
speed up leave approbation

speed up leave approbation by only recomputing the forecast lines in the
period of the leave

gurneyalex committed:
reduce log verbosity

gurneyalex committed:
refactor SaleOrder tests

gurneyalex committed:
improve unlink() management

use ondelete='cascade' to handle the unlinking of forecast.line records

ntsirintanis committed:
[UPD] cherry-pick and corrections from 15.0, pre-commit, tests

do not mix forecast.role, hr.employee.forecast.role when assigning an hr.job to an hr.employee,

Fixed bug by adding hr.group_hr_user group to field not available in hr.employee.public,

loop protection against duplicate forecast line creation

holidays test: somehow hr.leave.action_validate does not trigger hr.employee.forecast.role forecast line recreation, add an explicit trigger for this. Also add 2x loop protection on write

gurneyalex committed:
[FIX] forecast_line: issue in multicompany

when running in multi company, sometimes the user making a change is not connected
using the company of the record. In this case the default value for company_id on
forecast.line will lead to a wrong forecast.

we fix this by using the best company when updating the forecast (the
employee's company for task assigned to employees or leaves, the sale
line's company for sales forecast

gurneyalex committed:
[FIX] wrong condition in task quick update

the condition was wrong, preventing the fast update from being used
when only the remaining time is updated

gurneyalex committed:
[FIX] crash when a task is not related to a project

gurneyalex committed:
fixup! [FIX] forecast_line: issue in multicompany

gurneyalex committed:
[FIX] quick update sign inversion

when using the quick update, we would get an inversion of the sign of the forecast
because the ratio was negative

thomaspaulb commented:
@ntsirintanis Could you rebase once more and we see how the tests run now

thomaspaulb commented on project_forecast_line/tests/test_forecast_line.py:
```diff
@@ -0,0 +1,875 @@
+        self.test_task_forecast_lines_consolidated_forecast()
+        with Form(self.env["hr.leave"]) as form:
+            form.employee_id = self.employee_consultant
+            form.holiday_status_id = self.env.ref("hr_holidays.holiday_status_unpaid")
```
@ntsirintanis Currently this test is failing on `odoo.exceptions.ValidationError: Unpaid are only valid starting from 2023-01-01`.

I think it's because of [this line](https://github.com/odoo/odoo/blob/14.0/addons/hr_holidays/data/hr_holidays_data.xml#L49), which is removed in Odoo 15.0.

I think we can patch this test right at the beginning by removing or changing `validity_start` of `hr_holidays.holiday_status_unpaid`?

Commit:
[FIX] Fixed bug by adding hr.group_hr_user group to field not available in hr.employee.public

thomaspaulb commented on project_forecast_line/models/hr_employee.py:
```diff
@@ -0,0 +1,147 @@
+    role_ids = fields.One2many("hr.employee.forecast.role", "employee_id")
+    main_role_id = fields.Many2one(
+        "forecast.role", compute="_compute_main_role_id", ondelete="restrict"
+    )
```
Please add backport / cherry-pick of https://github.com/OCA/project/pull/1145

gjotten commented:
@thomaspaulb remaining problems in CI here are pre-commit as well as big failures in tests. Is this something we can bother you with?

gjotten reviewed (approved):
I can give a functional green light here by the way. What's included in the original 15.0 version works for 14.0 as well.

thomaspaulb commented:
Hm, the remaining failing test, I think, is an interaction with `project_recalculate`, which doesn't exist in 15.0.

I can't put my finger behind it yet, though

gjotten commented:
@thomaspaulb you feel like giving this a thumbs-up yourself as well? Or would that be a moral gray area 😄

Commit:
[FIX] project_timesheet_time_control: make tests timezone-resilient

thomaspaulb commented:
Well I've finally managed to make the tests green, but:

- [ ] It still needs some Git love, the commit order is now all jambled. We should rebase on latest 15.0, cherry-pick in #1106 and #1145 and then our own commits to backport should come
- [ ] Of our own commits, some, but not all, should be squashed. The loop protection may have to find its way to 15.0 as well, so could be kept separate; and the test-improvement in the other module can be separate as well.
- [ ] Afterwards, it would be really nice if @gurneyalex or @yankinmax could review again as well.

thomaspaulb commented:
@ntsirintanis Do you feel up to the Git challenge?

ntsirintanis commented:
@thomaspaulb I squashed something like 10 commits away, will revisit if more squashing is needed.

ntsirintanis committed:
[14.0][FIX] introduce employee_company_id field in hr.leave

ntsirintanis committed:
[UPD] project_forecast_line: rearrange commits and fix errors that occured

gfcapalbo committed:
[FIX] mixin should be Abstract model

gjotten commented:
@OCA/project-service-maintainers 'no stale' should be applied here.

@ntsirintanis I believe you looked into fixing some shortcuts we made here?

ntsirintanis commented:
@gjotten it's been a while since I pushed anything here, I'll see into making this specific none green soon. And if there are new commits for the module. It looks like that I just need to extend some tests

leemannd commented:
Hello @ntsirintanis , @gjotten ,
I added the label `no stale`.

I don't know much these modules. From what I understand. You still want to improve unit tests before merging?

ntsirintanis commented:
> Hello @ntsirintanis , @gjotten , I added the label `no stale`.
> 
> I don't know much these modules. From what I understand. You still want to improve unit tests before merging?

Thanks @leemannd . Yes. The purpose is to make this green and mergeable, ultimately. We have been using it for a while so it is quite stable.

leemannd commented on project_forecast_line/__manifest__.py:
```diff
@@ -0,0 +1,38 @@
+        "demo/sale.xml",
+    ],
+    "installable": True,
+    "development_status": "Alpha",
```
Open question, as you're using it since  while. Do you think this module is still in `Alpha` ?

gjotten replied:
As its sole user internally, I would say yes for the 14.0 version. Also, where we do the backport from in 15.0 also considers this to be in alpha.

yankinmax commented on project_forecast_line/models/hr_employee.py:
```diff
@@ -0,0 +1,189 @@
+        # can"t store as it depends on current date
+        today = fields.Date.context_today(self)
+        for rec in self:
+            rec.main_role_id = rec.role_ids.filtered(
```
let's get a first (or only one) record from recordset with `fields.first`

ntsirintanis committed:
[UPD] unit test coverage, remove create(), write() in project.task

ntsirintanis commented:
alright, this is green again

ntsirintanis commented:
> alright, this is green again

no, I was too hasty to comment

ntsirintanis commented:
ΟΚ, now truly this one is green

leemannd commented:
/ocabot merge nobump

ntsirintanis commented:
@leemannd thank you very very much for merging this. Can you please reopen this one https://github.com/OCA/project/pull/976 so we can get it merged too?
